File: packages/aid-rs/src/pka.rs
//! PKA handshake verification for Rust (feature = "handshake")
#![cfg(feature = "handshake")]

use crate::errors::AidError;
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine as _;
use ed25519_dalek::{Signature, VerifyingKey};
use httpdate::parse_http_date;
use reqwest::header::HeaderMap;
use reqwest::Client;
use std::collections::HashSet;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

fn multibase_decode(input: &str) -> Result<Vec<u8>, AidError> {
    if input.is_empty() {
        return Err(AidError::new("ERR_SECURITY", "Empty PKA"));
    }
    let (prefix, rest) = input.split_at(1);
    match prefix {
        "z" => bs58::decode(rest)
            .into_vec()
            .map_err(|_| AidError::new("ERR_SECURITY", "Invalid base58")),
        _ => Err(AidError::new("ERR_SECURITY", "Unsupported multibase prefix")),
    }
}

fn parse_signature_headers(headers: &HeaderMap) -> Result<(Vec<String>, i64, String, String, Vec<u8>, Option<String>), AidError> {
    let sig_input = headers
        .get("Signature-Input")
        .or_else(|| headers.get("signature-input"))
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| AidError::new("ERR_SECURITY", "Missing signature headers"))?;
    let sig = headers
        .get("Signature")
        .or_else(|| headers.get("signature"))
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| AidError::new("ERR_SECURITY", "Missing signature headers"))?;

    // Extract covered fields inside parentheses after sig=(...)
    let inside_start = sig_input.find("sig=(").ok_or_else(|| AidError::new("ERR_SECURITY", "Invalid Signature-Input"))? + 5;
    let rest = &sig_input[inside_start..];
    let close = rest.find(')').ok_or_else(|| AidError::new("ERR_SECURITY", "Invalid Signature-Input"))?;
    let inside = &rest[..close];
    let mut covered = Vec::new();
    let mut s = inside;
    while let Some(i) = s.find('"') {
        s = &s[i + 1..];
        if let Some(j) = s.find('"') {
            covered.push(s[..j].to_string());
            s = &s[j + 1..];
        } else {
            break;
        }
    }
    if covered.is_empty() {
        return Err(AidError::new("ERR_SECURITY", "Invalid Signature-Input"));
    }
    let required: HashSet<&str> = ["aid-challenge", "@method", "@target-uri", "host", "date"].into_iter().collect();
    let lower: HashSet<String> = covered.iter().map(|c| c.to_lowercase()).collect();
    if lower.len() != required.len() || required.iter().any(|r| !lower.contains(&r.to_string())) {
        return Err(AidError::new("ERR_SECURITY", "Signature-Input must cover required fields"));
    }

    // Params
    let mut created: i64 = 0;
    let mut keyid = String::new();
    let mut alg = String::new();
    for part in sig_input.split(';') {
        let p = part.trim();
        let pl = p.to_lowercase();
        if pl.starts_with("created=") {
            if let Ok(c) = p[8..].parse::<i64>() {
                created = c;
            }
        } else if pl.starts_with("keyid=") {
            keyid = p[6..].trim().to_string();
        } else if pl.starts_with("alg=") {
            alg = p[4..].trim().trim_matches('"').to_lowercase();
        }
    }
    if created == 0 || keyid.is_empty() || alg.is_empty() {
        return Err(AidError::new("ERR_SECURITY", "Invalid Signature-Input"));
    }
    // Signature header: sig=:base64:
    let sig_pos = sig.to_lowercase().find("sig=").ok_or_else(|| AidError::new("ERR_SECURITY", "Invalid Signature header"))?;
    let val = &sig[sig_pos + 4..];
    let val = val.strip_prefix(':').ok_or_else(|| AidError::new("ERR_SECURITY", "Invalid Signature header"))?;
    let end = val.find(':').ok_or_else(|| AidError::new("ERR_SECURITY", "Invalid Signature header"))?;
    let b64 = &val[..end];
    let signature = B64.decode(b64).map_err(|_| AidError::new("ERR_SECURITY", "Invalid Signature header"))?;

    let response_date = headers
        .get("Date")
        .or_else(|| headers.get("date"))
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string());
    Ok((covered, created, keyid, alg, signature, response_date))
}

fn build_signature_base(
    covered: &[String],
    created: i64,
    keyid: &str,
    alg: &str,
    method: &str,
    target_uri: &str,
    host: &str,
    date: &str,
    challenge: &str,
) -> Vec<u8> {
    let mut lines: Vec<String> = Vec::new();
    for item in covered {
        match item.to_lowercase().as_str() {
            "aid-challenge" => lines.push(format!("\"AID-Challenge\": {}", challenge)),
            "@method" => lines.push(format!("\"@method\": {}", method)),
            "@target-uri" => lines.push(format!("\"@target-uri\": {}", target_uri)),
            "host" => lines.push(format!("\"host\": {}", host)),
            "date" => lines.push(format!("\"date\": {}", date)),
            _ => return Vec::new(),
        }
    }
    let quoted = covered.iter().map(|c| format!("\"{}\"", c)).collect::<Vec<_>>().join(" ");
    let params = format!("({});created={};keyid={};alg=\"{}\"", quoted, created, keyid, alg);
    lines.push(format!("\"@signature-params\": {}", params));
    lines.join("\n").into_bytes()
}

pub async fn perform_pka_handshake(uri: &str, pka: &str, kid: &str, timeout: Duration) -> Result<(), AidError> {
    if kid.is_empty() {
        return Err(AidError::new("ERR_SECURITY", "Missing kid for PKA"));
    }
    let u = reqwest::Url::parse(uri).map_err(|_| AidError::new("ERR_SECURITY", "Invalid URI for handshake"))?;
    let host = u.host_str().ok_or_else(|| AidError::new("ERR_SECURITY", "Invalid URI for handshake"))?;
    let client = Client::builder().http2_prior_knowledge(false).timeout(timeout).build().map_err(|e| AidError::new("ERR_SECURITY", e.to_string()))?;

    // 32-byte random challenge; here we use system time jitter for simplicity; production should use rng
    // but reqwest + std provides no RNG-free API; however challenge is not secret
    let nonce = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos().to_le_bytes();
    let challenge = base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(&nonce);
    let date = httpdate::fmt_http_date(SystemTime::now());

    let res = client
        .get(u.clone())
        .header("AID-Challenge", &challenge)
        .header("Date", &date)
        .send()
        .await
        .map_err(|e| AidError::new("ERR_SECURITY", e.to_string()))?;
    if !res.status().is_success() {
        return Err(AidError::new("ERR_SECURITY", format!("Handshake HTTP {}", res.status())));
    }
    let headers = res.headers().clone();
    let (covered, created, mut keyid, alg, signature, response_date) = parse_signature_headers(&headers)?;
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
    if (now - created).abs() > 300 {
        return Err(AidError::new("ERR_SECURITY", "Signature created timestamp outside acceptance window"));
    }
    if let Some(date_hdr) = response_date {
        if let Ok(dt) = parse_http_date(&date_hdr) {
            let epoch = dt.duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
            if (now - epoch).abs() > 300 {
                return Err(AidError::new("ERR_SECURITY", "HTTP Date header outside acceptance window"));
            }
        } else {
            return Err(AidError::new("ERR_SECURITY", "Invalid Date header"));
        }
    }
    if keyid.len() >= 2 && keyid.starts_with('"') && keyid.ends_with('"') {
        keyid = keyid.trim_matches('"').to_string();
    }
    if keyid != kid {
        return Err(AidError::new("ERR_SECURITY", "Signature keyid mismatch"));
    }
    if alg.to_lowercase() != "ed25519" {
        return Err(AidError::new("ERR_SECURITY", "Unsupported signature algorithm"));
    }

    let base = build_signature_base(
        &covered,
        created,
        &keyid,
        &alg,
        "GET",
        uri,
        host,
        response_date.as_deref().unwrap_or(&date),
        &challenge,
    );

    let pubkey = multibase_decode(pka)?;
    if pubkey.len() != 32 {
        return Err(AidError::new("ERR_SECURITY", "Invalid PKA length"));
    }
    let vk = VerifyingKey::from_bytes(pubkey.as_slice().try_into().unwrap())
        .map_err(|_| AidError::new("ERR_SECURITY", "Invalid public key"))?;
    let sig = Signature::from_slice(&signature).map_err(|_| AidError::new("ERR_SECURITY", "Invalid signature"))?;
    vk.verify(&base, &sig)
        .map_err(|_| AidError::new("ERR_SECURITY", "PKA signature verification failed"))?;
    Ok(())
}

File: packages/aid/src/pka.ts
import { AidError } from './parser.js';
import { webcrypto as nodeWebcrypto } from 'node:crypto';
import { Buffer } from 'node:buffer';
import { timingSafeEqual as nodeTimingSafeEqual } from 'node:crypto';

// Type-safe interface for global crypto with timingSafeEqual
interface CryptoWithTimingSafeEqual {
  timingSafeEqual(a: Uint8Array, b: Uint8Array): boolean;
}

function timingSafeEqual(a: string | Uint8Array, b: string | Uint8Array): boolean {
  const globalCrypto = (globalThis as unknown as { crypto?: CryptoWithTimingSafeEqual }).crypto;
  if (typeof globalCrypto?.timingSafeEqual === 'function') {
    const enc = new TextEncoder();
    const aEncoded = typeof a === 'string' ? enc.encode(a) : a;
    const bEncoded = typeof b === 'string' ? enc.encode(b) : b;
    return globalCrypto.timingSafeEqual(aEncoded, bEncoded);
  }
  // Fallback for environments without native support, including browsers.
  const aBuf = Buffer.from(typeof a === 'string' ? a : new Uint8Array(a.buffer));
  const bBuf = Buffer.from(typeof b === 'string' ? b : new Uint8Array(b.buffer));

  if (aBuf.length !== bBuf.length) {
    // For string comparisons, we must ensure they are of equal length.
    // In this PKA context, `keyid` and `alg` have predictable lengths,
    // so length differences don't leak critical info. We still check
    // b against itself to keep the timing consistent.
    nodeTimingSafeEqual(bBuf, bBuf);
    return false;
  }
  return nodeTimingSafeEqual(aBuf, bBuf);
}

function asciiLowerCase(s: string): string {
  let res = '';
  for (let i = 0; i < s.length; i++) {
    const charCode = s.charCodeAt(i);
    // ASCII 'A' is 65, 'Z' is 90
    if (charCode >= 65 && charCode <= 90) {
      res += String.fromCharCode(charCode + 32);
    } else {
      res += s[i];
    }
  }
  return res;
}

// Minimal types to avoid DOM deps
interface HeaderLike {
  get(name: string): string | null | undefined;
}

interface SubtleLike {
  importKey: (
    format: 'raw',
    keyData: ArrayBuffer | Uint8Array,
    algorithm: { name: 'Ed25519' },
    extractable: false,
    keyUsages: ['verify'],
  ) => Promise<unknown>;
  verify: (
    algorithm: 'Ed25519',
    key: unknown,
    signature: ArrayBufferView,
    data: ArrayBufferView,
  ) => Promise<boolean>;
}

interface CryptoLike {
  getRandomValues: (array: Uint8Array) => Uint8Array;
  subtle: SubtleLike;
}

interface FetchResponse {
  ok: boolean;
  status: number;
  headers: HeaderLike;
  text(): Promise<string>;
}

type FetchLike = (
  url: string,
  init: {
    method?: string;
    headers?: Record<string, string>;
    redirect?: 'error' | 'follow' | 'manual';
    signal?: AbortSignal;
  },
) => Promise<FetchResponse>;

// Base58btc alphabet
const B58 = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
const B58_MAP = new Map<string, number>(Array.from(B58).map((c, i) => [c, i]));

function base58Decode(s: string): Uint8Array {
  if (!s) return new Uint8Array();
  let zeros = 0;
  while (zeros < s.length && s[zeros] === '1') zeros++;
  const size = (((s.length - zeros) * Math.log(58)) / Math.log(256) + 1) | 0;
  const b = new Uint8Array(size);
  for (let i = zeros; i < s.length; i++) {
    const c = s[i];
    const val = B58_MAP.get(c);
    if (val === undefined) throw new AidError('ERR_SECURITY', 'Invalid base58 character');
    let carry = val;
    for (let j = size - 1; j >= 0; j--) {
      carry += 58 * b[j];
      b[j] = carry & 0xff;
      carry >>= 8;
    }
  }
  // Skip leading zeros in b
  let it = 0;
  while (it < b.length && b[it] === 0) it++;
  const out = new Uint8Array(zeros + (b.length - it));
  out.fill(0, 0, zeros);
  out.set(b.subarray(it), zeros);
  return out;
}

function multibaseDecode(input: string): Uint8Array {
  if (!input) throw new AidError('ERR_SECURITY', 'Empty PKA');
  const prefix = input[0];
  const payload = input.slice(1);
  if (prefix === 'z') {
    return base58Decode(payload);
  }
  throw new AidError('ERR_SECURITY', 'Unsupported multibase prefix');
}

function parseSignatureHeaders(headers: HeaderLike): {
  covered: string[];
  created: number;
  keyid: string; // normalized (no quotes)
  keyidRaw: string; // as present in Signature-Input (may include quotes)
  alg: string;
  signature: Uint8Array;
  responseDate: string | null;
} {
  const sigInput = headers.get('Signature-Input') || headers.get('signature-input');
  const sig = headers.get('Signature') || headers.get('signature');
  if (!sigInput || !sig) throw new AidError('ERR_SECURITY', 'Missing signature headers');

  // Extract covered fields inside parentheses after sig=(...)
  const inside = /sig=\(\s*([^)]*?)\s*\)/i.exec(sigInput);
  if (!inside) throw new AidError('ERR_SECURITY', 'Invalid Signature-Input');
  const covered: string[] = [];
  const tokenRe = /"([^"]+)"/g;
  let m: RegExpExecArray | null;
  while ((m = tokenRe.exec(inside[1])) !== null) covered.push(m[1]);
  if (covered.length === 0) throw new AidError('ERR_SECURITY', 'Invalid Signature-Input');
  const required = ['aid-challenge', '@method', '@target-uri', 'host', 'date'];

  if (covered.length !== required.length) {
    throw new AidError('ERR_SECURITY', 'Signature-Input must cover required fields');
  }

  const coveredLower = covered.map(asciiLowerCase).sort();
  const requiredSorted = [...required].sort();

  let areEqual = true;
  for (let i = 0; i < requiredSorted.length; i++) {
    if (!timingSafeEqual(coveredLower[i], requiredSorted[i])) {
      areEqual = false;
      // Do not break early
    }
  }
  if (!areEqual) {
    throw new AidError('ERR_SECURITY', 'Signature-Input must cover required fields');
  }

  // Extract parameters regardless of order
  const createdMatch = /(?:^|;)\s*created=(\d+)/i.exec(sigInput);
  const keyidMatch = /(?:^|;)\s*keyid=([^;\s]+)/i.exec(sigInput);
  const algMatch = /(?:^|;)\s*alg="([^"]+)"/i.exec(sigInput);
  if (!createdMatch || !keyidMatch || !algMatch)
    throw new AidError('ERR_SECURITY', 'Invalid Signature-Input');
  const created = Number.parseInt(createdMatch[1], 10);
  const keyidRaw = keyidMatch[1];
  const keyid = keyidRaw.replace(/^"(.+)"$/, '$1');
  const alg = asciiLowerCase(algMatch[1]);

  // Extract signature value from Signature header
  const sigMatch = /sig\s*=\s*:\s*([^:]+)\s*:/i.exec(sig);
  if (!sigMatch) throw new AidError('ERR_SECURITY', 'Invalid Signature header');
  const signature = Uint8Array.from(Buffer.from(sigMatch[1], 'base64'));
  const responseDate = (headers.get('Date') || headers.get('date') || null) as string | null;
  return { covered, created, keyid, keyidRaw, alg, signature, responseDate };
}

function buildSignatureBase(
  covered: string[],
  params: { created: number; keyid: string; alg: string },
  ctx: {
    method: string;
    targetUri: string;
    host: string;
    date: string;
    challenge: string;
  },
): Uint8Array {
  const lines: string[] = [];
  for (const item of covered) {
    const lower = asciiLowerCase(item);
    let appended = false;
    if (timingSafeEqual(lower, 'aid-challenge')) {
      lines.push(`"AID-Challenge": ${ctx.challenge}`);
      appended = true;
    }
    if (timingSafeEqual(lower, '@method')) {
      lines.push(`"@method": ${ctx.method}`);
      appended = true;
    }
    if (timingSafeEqual(lower, '@target-uri')) {
      lines.push(`"@target-uri": ${ctx.targetUri}`);
      appended = true;
    }
    if (timingSafeEqual(lower, 'host')) {
      lines.push(`"host": ${ctx.host}`);
      appended = true;
    }
    if (timingSafeEqual(lower, 'date')) {
      lines.push(`"date": ${ctx.date}`);
      appended = true;
    }
    if (!appended) {
      throw new AidError('ERR_SECURITY', `Unsupported covered field: ${item}`);
    }
  }
  const quoted = covered.map((c) => `"${c}"`).join(' ');
  const paramsStr = `(${quoted});created=${params.created};keyid=${params.keyid};alg="${params.alg}"`;
  lines.push(`"@signature-params": ${paramsStr}`);
  return new TextEncoder().encode(lines.join('\n'));
}

export async function performPKAHandshake(uri: string, pka: string, kid: string): Promise<void> {
  if (!kid) throw new AidError('ERR_SECURITY', 'Missing kid for PKA');
  const u = new URL(uri);
  const cryptoImpl: CryptoLike =
    (globalThis as unknown as { crypto?: CryptoLike }).crypto ??
    (nodeWebcrypto as unknown as CryptoLike);
  const nonce = cryptoImpl.getRandomValues(new Uint8Array(32));
  const challenge = Buffer.from(nonce).toString('base64url');
  const date = new Date().toUTCString();
  const fetchImpl = (globalThis as unknown as { fetch?: FetchLike }).fetch;
  if (typeof fetchImpl !== 'function') {
    throw new AidError('ERR_SECURITY', 'fetch is not available in this environment');
  }
  const res = await fetchImpl(uri, {
    method: 'GET',
    headers: {
      'AID-Challenge': challenge,
      Date: date,
    },
    redirect: 'error',
  });
  if (!res.ok) throw new AidError('ERR_SECURITY', `Handshake HTTP ${res.status}`);

  const { covered, created, keyid, keyidRaw, alg, signature, responseDate } = parseSignatureHeaders(
    res.headers,
  );
  const now = Math.floor(Date.now() / 1000);
  if (Math.abs(now - created) > 300)
    throw new AidError('ERR_SECURITY', 'Signature created timestamp outside acceptance window');
  if (responseDate) {
    const parsed = Math.floor(new Date(responseDate).getTime() / 1000);
    if (!Number.isFinite(parsed)) throw new AidError('ERR_SECURITY', 'Invalid Date header');
    if (Math.abs(now - parsed) > 300)
      throw new AidError('ERR_SECURITY', 'HTTP Date header outside acceptance window');
  }
  if (!timingSafeEqual(keyid, kid)) throw new AidError('ERR_SECURITY', 'Signature keyid mismatch');
  if (!timingSafeEqual(alg, 'ed25519'))
    throw new AidError('ERR_SECURITY', 'Unsupported signature algorithm');

  const host = u.host;
  const base = buildSignatureBase(
    covered,
    { created, keyid: keyidRaw, alg },
    {
      method: 'GET',
      targetUri: uri,
      host,
      date: responseDate ?? date,
      challenge,
    },
  );
  const pub = multibaseDecode(pka);
  if (pub.length !== 32) throw new AidError('ERR_SECURITY', 'Invalid PKA length');

  const key = await cryptoImpl.subtle.importKey('raw', pub, { name: 'Ed25519' }, false, ['verify']);
  const ok = await cryptoImpl.subtle.verify('Ed25519', key, signature, base);
  if (!ok) throw new AidError('ERR_SECURITY', 'PKA signature verification failed');
}
